wom/src/main/scala/wdl4s/wdl/WdlGraphNode.scala

```diff
@@ -46,22 +53,64 @@ object WdlGraphNode {
     updatesNeeded.foldLeft(setWithUpstream)(calculateUpstreamAncestry)
   }
 
-  private[wdl] def outputPortFromNode(node: WdlGraphNode, terminal: Option[Terminal]): ErrorOr[GraphNodePort.OutputPort] = {
+  /**
+    * Build a Wom Graph from a Scope in WDL.
+    * - Looks at the children of the Scope and converts them into GraphNodes.
+    * - Adds in any extras from 'includeGraphNode' which the call knows should also be in the Graph (eg the Scatter variable's InputGraphNode which this method doesn't know exists)
+    * - Builds them all together into a Graph
+    */
+  private[wdl] def buildWomGraph(from: Scope, includeGraphNodes: Set[GraphNode], outerLookup: Map[String, GraphNodePort.OutputPort]): ErrorOr[Graph] = {
+
+    final case class FoldState(nodes: Set[GraphNode], availableInputs: Map[String, GraphNodePort.OutputPort])
+
+    val initialFoldState = FoldState(
+      Set.empty,
+      includeGraphNodes.flatMap { gn =>
+        gn.outputPorts map { p => p.name -> p }
+      }.toMap
+    )
+
+    def foldFunction(acc: ErrorOr[FoldState], node: WdlGraphNode): ErrorOr[FoldState] = acc flatMap { goodAcc =>  buildNode(goodAcc, node) }
```

Comment:
Question for cat-people, I suspect there's probably a clever method called something like `foldTraverse` that'll handle this for me?

Reply:
I need to digest this chunk and I'll get back to you...